### ISSUE TITLE
sql/kv: support blind writes for UPDATE statements

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#71153]
Currently, the evaluation of a single-row UPDATE statement results in a `Get` KV request followed by a `Put` KV request. We can see this in the following trace:
```sql
root@127.0.0.1:26257/defaultdb> CREATE TABLE kv (k INT PRIMARY KEY, v INT);
CREATE TABLE

root@127.0.0.1:26257/defaultdb> INSERT INTO kv VALUES (1, 2);
INSERT 1

root@127.0.0.1:26257/defaultdb> SET TRACING = ON; UPDATE kv SET v = 3 WHERE k = 1; SET TRACING = OFF;
SET TRACING

root@127.0.0.1:26257/defaultdb> SHOW KV TRACE FOR SESSION;
            timestamp           |       age       |                     message                     |                              tag                              |                location                 |       operation        | span
--------------------------------+-----------------+-------------------------------------------------+---------------------------------------------------------------+-----------------------------------------+------------------------+-------
  2021-10-05 16:06:54.112223+00 | 00:00:00.000987 | Scan /Table/59/1/1/0                            | [n=1,client=127.0.0.1:51280,hostnossl,user=root]              | sql/row/kv_batch_fetcher.go:426         | colbatchscan           |    7
  2021-10-05 16:06:54.112279+00 | 00:00:00.001043 | querying next range at /Table/59/1/1/0          | [n=1,client=127.0.0.1:51280,hostnossl,user=root,txn=fa940497] | kv/kvclient/kvcoord/range_iter.go:175   | dist sender send       |    9
  2021-10-05 16:06:54.112315+00 | 00:00:00.001079 | r74: sending batch 1 Get to (n1,s1):1           | [n=1,client=127.0.0.1:51280,hostnossl,user=root,txn=fa940497] | kv/kvclient/kvcoord/dist_sender.go:1930 | dist sender send       |    9
  2021-10-05 16:06:54.112861+00 | 00:00:00.001625 | querying next range at /Table/59/1/1/0          | [n=1,client=127.0.0.1:51280,hostnossl,user=root,txn=fa940497] | kv/kvclient/kvcoord/range_iter.go:175   | dist sender send       |   12
  2021-10-05 16:06:54.112885+00 | 00:00:00.001649 | r74: sending batch 1 Put, 1 EndTxn to (n1,s1):1 | [n=1,client=127.0.0.1:51280,hostnossl,user=root,txn=fa940497] | kv/kvclient/kvcoord/dist_sender.go:1930 | dist sender send       |   12
  2021-10-05 16:06:54.115285+00 | 00:00:00.004049 | fast path completed                             | [n=1,client=127.0.0.1:51280,hostnossl,user=root]              | sql/plan_node_to_row_source.go:159      | batch flow coordinator |    6
  2021-10-05 16:06:54.115488+00 | 00:00:00.004252 | rows affected: 1                                | [n=1,client=127.0.0.1:51280,hostnossl,user=root]              | sql/conn_executor_exec.go:713           | exec stmt              |    3
(7 rows)
```

The `Get` request serves two purposes. First, it determines whether a matching row exists to determine whether a write is needed or not. If not, UPDATE should not perform a write and should return 0 rows affected. Second, it retrieves the existing columns in the row. SQL may need the existing columns either because they are used in the UPDATE's SET clause or because they are not included in the UPDATE and need to be written back to KV as part of the larger column family (implicit `new = prev` set clause).

However, the `Get` request comes at a cost. Primarily, it requires a network hop, a disk read, and some additional processing. This also has a secondary cost that has to do with contention. The separate `Get` and `Put` require CRDB to acquire an unreplicated lock during the read and prevent the `Put` from performing a server-side refresh if its timestamp gets bumped. This all contributes to the difference in execution performance between `UPDATE` statements and `UPSERT` statements (which can perform a blind write in some cases).

We should add a fast-path for blind writes (i.e. a `Put` without the `Get`) for a subset of `UPDATE` statements where this is possible.

#### KV

KV will introduce a new form of `ConditionalPut` request whose condition matches with any non-tombstone value. For the sake of this issue, consider this `ConditionalPut(exp=*)`. This will allow SQL to issue a blind write that returns whether or not the target key exists and does nothing if the key does not exist.

KV will also introduce an option on `ConditionalPut` to have it return a boolean indicating whether it wrote or not instead of an error, to avoid putting the txn in an error state if it does not write. Care will need to be taken with the timestamp cache and with pipelining (ref https://github.com/cockroachdb/cockroach/issues/64723) when we introduce this option.

#### SQL

SQL will introduce a new fast-path that issues a `ConditionalPut(exp=*)` for an UPDATE statement where all columns in the row and/or column families being touched are known without reading the row. This fast-path will skip the initial row fetch.

### Impact

I hypothesize that this would result in a sizable improvement to YCSB A and B, somewhere around 30%. An interesting next step would be to prototype this change and check. Or we could run YCSB with `UPSERT` statements instead of `UPDATE` statements to get a feel for how much the locking `Get` is costing us. If we do that, we'll need to tweak the behavior of `UPSERT` with multiple column families slightly, because we only want to touch one column family.

**Comment by mgartner** [comment]
@nvanbenschoten I'm assuming that KV has not yet implemented the `ConditionalPut` mentioned above, correct? Should we add this issue to the KV project as well?

**Comment by nvanbenschoten** [comment]
> I'm assuming that KV has not yet implemented the ConditionalPut mentioned above, correct?

That's correct. I'll add it to the kv project as well.

I this something we are considering doing?

**Comment by mgartner** [comment]
> I this something we are considering doing?

No, just grooming the backlog. This optimization would only apply to a very narrow `UPDATE` and table schema, so probably not worth it until we have a very strong demand.

**Comment by michae2** [comment]
It seems like a limited form of this could be accomplished today using the existing `CPut` if the UPDATE statement were modified from:

```sql
UPDATE kv SET v = 3 WHERE k = 1;
```

to:

```sql
UPDATE kv SET v = 3 WHERE k = 1 AND v = 2;
```
